Bug 1886272: Starting suggest benchmarks

bendk opened the pull request:
These can be run using `cargo suggest-bench`.
The only benchmark I added so far is the time for a full ingestion.

### Pull Request checklist ###
<!-- Before submitting the PR, please address each item -->
- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [ ] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

bendk commented:
On my machine:
 - Ingestion takes about 1.5 seconds
 - The time to download all the remote settings data is very similar.

I think this means that:
  - We should definitely be able to speed up ingestion
  - I'm not sure if this will fix the firefox-ios issue though, since the network requests themselves may cause the timeout.

ncloudioj commented on components/suggest/benches/benchmark_all.rs:
```diff
@@ -0,0 +1,19 @@
+pub fn ingest_benchmarks(c: &mut Criterion) {
+    let mut group = c.benchmark_group("ingest");
+    viaduct_reqwest::use_reqwest_backend();
+    group.sample_size(10);
```
10 is the minimal sample size, any particular reason we want to pin it?

bendk replied:
The default of 100 makes the tests take a long time.  I think they would take around 150s, but I'm not totally sure because I got tired of waiting and killed them.  I'll add a note about this.

ncloudioj commented on components/suggest/src/benchmarks/client.rs:
```diff
@@ -0,0 +1,97 @@
+/// Remotes settings client that runs during the benchmark warm-up phase.
+///
+/// This should be used to run a full ingestion.
+/// Then in can be converted into a [RemoteSettingsBenchmarkClient], which helps allows benchmark code to exclude the network request time.
```
```suggestion
/// Then iit can be converted into a [RemoteSettingsBenchmarkClient], which allows benchmark code to exclude the network request time.
```

ncloudioj reviewed (approved):
This looks great, thanks!

bendk commented:
I made a few changes based on our conversation and also from me trying to actually use these benchmarks:
  - Added a README.md file which describes the benchmarks and also lists some missing benchmarks that would be nice to have.
  - Split up the benchmark by record type.  Most record types take a small amount of time, it's really only AMP-wikipedia and AMP-mobile that are driving the total time.  I don't think we lose anything by splitting up the benchmarks, since when I add those 2 times together, I get approximately the same time as the total.

ncloudioj commented on components/suggest/src/benchmarks/client.rs:
```diff
@@ -0,0 +1,97 @@
+/// Remotes settings client that runs during the benchmark warm-up phase.
+///
+/// This should be used to run a full ingestion.
+/// Then iit can be converted into a [RemoteSettingsBenchmarkClient], which allows benchmark code to exclude the network request time.
```
```suggestion
/// Then it can be converted into a [RemoteSettingsBenchmarkClient], which allows benchmark code to exclude the network request time.
```

ncloudioj reviewed (approved):
Like those changes, r+.
